internal/notification/handlers/notification_worker.go

```diff
@@ -0,0 +1,494 @@
+	Workers               uint8
+	BulkLimit             uint16
+	RequeueEvery          time.Duration
+	RetryWorkers          uint8
+	RetryRequeueEvery     time.Duration
+	HandleActiveInstances time.Duration
+	TransactionDuration   time.Duration
+	MaxAttempts           uint8
+	MaxTtl                time.Duration
+	MinRetryDelay         time.Duration
+	MaxRetryDelay         time.Duration
+	RetryDelayFactor      float32
+}
+
+// nowFunc makes [time.Now] mockable
+type nowFunc func() time.Time
+
+type Sent func(ctx context.Context, commands Commands, id, orgID string, generatorInfo *senders.CodeGeneratorInfo, args map[string]any) error
+
+var sentHandlers map[eventstore.EventType]Sent
+
+func RegisterSentHandler(eventType eventstore.EventType, sent Sent) {
+	if sentHandlers == nil {
+		sentHandlers = make(map[eventstore.EventType]Sent)
+	}
+	sentHandlers[eventType] = sent
+}
+
+func NewNotificationWorker(
+	config WorkerConfig,
+	commands Commands,
+	queries *NotificationQueries,
+	es *eventstore.Eventstore,
+	client *database.DB,
+	channels types.ChannelChains,
+) *NotificationWorker {
+	// make sure the delay does not get less
+	if config.RetryDelayFactor < 1 {
+		config.RetryDelayFactor = 1
+	}
+	w := &NotificationWorker{
+		config:   config,
+		commands: commands,
+		queries:  queries,
+		es:       es,
+		client:   client,
+		channels: channels,
+		now:      time.Now,
+	}
+	w.backOff = w.exponentialBackOff
+	return w
+}
+
+func (w *NotificationWorker) Start(ctx context.Context) {
+	for i := 0; i < int(w.config.Workers); i++ {
+		go w.schedule(ctx, i, false)
+	}
+	for i := 0; i < int(w.config.RetryWorkers); i++ {
+		go w.schedule(ctx, i, true)
+	}
+}
+
+func (w *NotificationWorker) reduceNotificationRequested(ctx context.Context, tx *sql.Tx, event *notification.RequestedEvent) (err error) {
+	ctx = ContextWithNotifier(ctx, event.Aggregate())
+
+	// if the notification is too old, we can directly cancel
+	if event.CreatedAt().Add(w.config.MaxTtl).Before(w.now()) {
+		return w.commands.NotificationCanceled(ctx, tx, event.Aggregate().ID, event.Aggregate().ResourceOwner, nil)
+	}
+
+	// Get the notify user first, so if anything fails afterward we have the current state of the user
+	// and can pass that to the retry request.
+	notifyUser, err := w.queries.GetNotifyUserByID(ctx, true, event.UserID)
+	if err != nil {
+		return err
+	}
+
+	// The domain claimed event requires the domain as argument, but lacks the user when creating the request event.
+	// Since we set it into the request arguments, it will be passed into a potential retry event.
+	if _, ok := event.Request.Args["Domain"]; !ok && event.RequiresPreviousDomain {
+		index := strings.LastIndex(notifyUser.LastEmail, "@")
+		event.Request.Args["Domain"] = notifyUser.LastEmail[index+1:]
+	}
+
+	err = w.sendNotification(ctx, tx, event.Request, notifyUser, event)
+	if err == nil {
+		return nil
+	}
+	// if retries are disabled, we cancel the notification
+	if w.config.MaxAttempts <= 1 {
```

Comment:
Wouldn't this mean that if you say MaxAttempts = 1, that you would never send it?